File: signals/context.py
from __future__ import annotations

import numpy as np
import pandas as pd

from data.symbols import benchmark_for_market
from signals.common import atr, efficiency_ratio, ema, obv, rolling_percentile
from signals.models import RegimeSignal, RelativeStrengthSignal, VolumeSignal


def analyze_volume_participation(frame: pd.DataFrame, breakout_level: float | None = None) -> VolumeSignal:
    working = frame.copy().sort_values("datetime").reset_index(drop=True)
    latest = working.iloc[-1]
    avg20 = working["volume"].tail(21).iloc[:-1].mean() if len(working) >= 21 else working["volume"].mean()
    volume_ratio = float(latest["volume"] / avg20) if avg20 else 1.0
    obv_series = obv(working)
    obv_confirmation = len(obv_series) > 5 and float(obv_series.iloc[-1]) >= float(obv_series.iloc[-5])
    typical_price = (working["high"] + working["low"] + working["close"]) / 3.0
    rolling_vwap = (typical_price * working["volume"]).rolling(20, min_periods=5).sum() / working["volume"].rolling(20, min_periods=5).sum()
    vwap_alignment = bool(pd.notna(rolling_vwap.iloc[-1]) and float(latest["close"]) >= float(rolling_vwap.iloc[-1]))
    anchor_idx = max(len(working) - 20, 0)
    if breakout_level is not None:
        matches = working.index[working["close"] >= breakout_level].tolist()
        if matches:
            anchor_idx = matches[0]
    anchored_slice = working.iloc[anchor_idx:]
    anchored_typical = (anchored_slice["high"] + anchored_slice["low"] + anchored_slice["close"]) / 3.0
    anchored_vwap = float((anchored_typical * anchored_slice["volume"]).sum() / anchored_slice["volume"].sum()) if anchored_slice["volume"].sum() else None
    dry_up = len(working) >= 11 and float(working["volume"].tail(5).mean()) < float(working["volume"].tail(20).mean()) * 0.85

    penalty_flags: list[str] = []
    if volume_ratio < 1.2:
        penalty_flags.append("breakout_without_participation")
    candle_spread = float(latest["high"] - latest["low"])
    atr_series = atr(working, 14)
    atr_value = float(atr_series.iloc[-1]) if pd.notna(atr_series.iloc[-1]) else 0.0
    if atr_value and candle_spread > atr_value * 1.5 and volume_ratio < 1.3:
        penalty_flags.append("wide_spread_weak_volume")
    if volume_ratio > 3.0 and float(latest["close"]) < float(latest["high"]) * 0.98:
        penalty_flags.append("possible_exhaustion")

    return VolumeSignal(
        volume_ratio=round(volume_ratio, 2),
        relative_volume=round(volume_ratio, 2),
        obv_confirmation=obv_confirmation,
        vwap_alignment=vwap_alignment,
        anchored_vwap=round(anchored_vwap, 4) if anchored_vwap is not None else None,
        dry_up_before_expansion=dry_up,
        penalty_flags=penalty_flags,
    )


def analyze_relative_strength(
    frame: pd.DataFrame,
    benchmark_frame: pd.DataFrame,
    *,
    benchmark_symbol: str,
    sector_frame: pd.DataFrame | None = None,
    sector_symbol: str | None = None,
) -> RelativeStrengthSignal:
    working = frame.copy().sort_values("datetime").reset_index(drop=True)
    benchmark = benchmark_frame.copy().sort_values("datetime").reset_index(drop=True)
    joined = pd.merge(
        working[["datetime", "close"]],
        benchmark[["datetime", "close"]],
        on="datetime",
        how="inner",
        suffixes=("_asset", "_benchmark"),
    )
    if joined.empty:
        return RelativeStrengthSignal(
            benchmark_symbol=benchmark_symbol,
            sector_benchmark_symbol=sector_symbol,
            score=0.0,
            trend_persistence=0.0,
            smoothness=0.0,
            one_week_alpha=0.0,
            one_month_alpha=0.0,
            three_month_alpha=0.0,
            explanation="Benchmark overlap was unavailable.",
        )

    def _alpha(window: int) -> float:
        if len(joined) < window + 1:
            return 0.0
        asset_return = float(joined["close_asset"].iloc[-1] / joined["close_asset"].iloc[-window - 1] - 1.0)
        benchmark_return = float(joined["close_benchmark"].iloc[-1] / joined["close_benchmark"].iloc[-window - 1] - 1.0)
        return asset_return - benchmark_return

    one_week = _alpha(5)
    one_month = _alpha(21)
    three_month = _alpha(63)
    rs_line = joined["close_asset"] / joined["close_benchmark"]
    smoothness = efficiency_ratio(rs_line, 20)
    persistence = float((rs_line.diff().tail(20) > 0).mean())
    raw_score = ((one_week * 20) + (one_month * 35) + (three_month * 45) + (smoothness * 20) + (persistence * 10)) * 10
    score = max(0.0, min(100.0, raw_score))
    return RelativeStrengthSignal(
        benchmark_symbol=benchmark_symbol,
        sector_benchmark_symbol=sector_symbol,
        score=round(score, 2),
        trend_persistence=round(persistence * 100.0, 2),
        smoothness=round(smoothness * 100.0, 2),
        one_week_alpha=round(one_week * 100.0, 2),
        one_month_alpha=round(one_month * 100.0, 2),
        three_month_alpha=round(three_month * 100.0, 2),
        explanation="Relative strength is measured against the selected benchmark with 1W/1M/3M alpha and path smoothness.",
    )


def analyze_market_regime(benchmark_frame: pd.DataFrame, *, market: str, benchmark_symbol: str) -> RegimeSignal:
    working = benchmark_frame.copy().sort_values("datetime").reset_index(drop=True)
    if working.empty or len(working) < 220:
        return RegimeSignal(
            market=market,
            benchmark_symbol=benchmark_symbol,
            direction="neutral",
            trend_strength=0.0,
            volatility_state="unknown",
            breadth_like_proxy=0.0,
            explanation="Insufficient benchmark history.",
        )
    working["ema20"] = ema(working["close"], 20)
    working["ema50"] = ema(working["close"], 50)
    working["ema200"] = ema(working["close"], 200)
    working["atr14"] = atr(working, 14)
    working["atr_pctile"] = rolling_percentile(working["atr14"], 252)
    latest = working.iloc[-1]
    bullish = float(latest["close"]) > float(latest["ema20"]) > float(latest["ema50"]) > float(latest["ema200"])
    bearish = float(latest["close"]) < float(latest["ema20"]) < float(latest["ema50"]) < float(latest["ema200"])
    direction = "bullish" if bullish else "bearish" if bearish else "neutral"
    trend_strength = float(
        np.mean(
            [
                float(float(latest["close"]) > float(latest["ema20"])),
                float(float(latest["ema20"]) > float(latest["ema50"])),
                float(float(latest["ema50"]) > float(latest["ema200"])),
            ],
        )
        * 100.0,
    )
    atr_pctile = float(latest["atr_pctile"]) if pd.notna(latest["atr_pctile"]) else 0.5
    volatility_state = "expanding" if atr_pctile > 0.65 else "compressed" if atr_pctile < 0.35 else "normal"
    breadth_proxy = float((working["close"].tail(50) > working["ema20"].tail(50)).mean() * 100.0)
    return RegimeSignal(
        market=market,
        benchmark_symbol=benchmark_symbol,
        direction=direction,
        trend_strength=round(trend_strength, 2),
        volatility_state=volatility_state,
        breadth_like_proxy=round(breadth_proxy, 2),
        explanation=f"{market} regime is {direction} with {volatility_state} volatility relative to {benchmark_symbol}.",
    )


def benchmark_symbol_for_market(market: str, benchmark_map: dict[str, dict[str, str]]) -> str:
    return benchmark_for_market(market, benchmark_map)

File: stocker/lib/engine/context.ts
/**
 * Context engine: volume participation, relative strength (with sector routing),
 * and market regime. Ported from STOPICK signals/context.py and extended with
 * an above-anchored-VWAP flag, an OBV accumulation trend, and sector alpha.
 */
import type {
  Bar,
  Direction,
  Market,
  RegimeSignal,
  RelativeStrengthSignal,
  VolumeSignal,
} from "./types";
import {
  atr,
  bounded,
  closes,
  ema,
  efficiencyRatio,
  last,
  linregSlope,
  mean,
  obv,
  rollingPercentile,
  round,
} from "./indicators";

export function analyzeVolumeParticipation(bars: Bar[], breakoutLevel: number | null = null): VolumeSignal {
  const latest = bars[bars.length - 1];
  const tail21 = bars.slice(-21);
  const avg20 = tail21.length >= 21 ? mean(tail21.slice(0, -1).map((b) => b.volume)) : mean(bars.map((b) => b.volume));
  const volumeRatio = avg20 ? latest.volume / avg20 : 1;

  const obvSeries = obv(bars);
  const obvConfirmation = obvSeries.length > 5 && obvSeries[obvSeries.length - 1] >= obvSeries[obvSeries.length - 6];
  const obvSlope = linregSlope(obvSeries.slice(-20), 20);
  const accumulationTrend = Math.sign(obvSlope);

  // Rolling 20-bar VWAP.
  const typical = bars.map((b) => (b.high + b.low + b.close) / 3);
  let vwapAlignment = false;
  {
    const w = 20;
    const slice = bars.slice(-w);
    if (slice.length >= 5) {
      const num = slice.reduce((a, b) => a + ((b.high + b.low + b.close) / 3) * b.volume, 0);
      const den = slice.reduce((a, b) => a + b.volume, 0);
      const vwap = den ? num / den : NaN;
      vwapAlignment = Number.isFinite(vwap) && latest.close >= vwap;
    }
  }

  // Anchored VWAP from the first close above the breakout level (else last 20 bars).
  let anchorIdx = Math.max(bars.length - 20, 0);
  if (breakoutLevel !== null) {
    const found = bars.findIndex((b) => b.close >= breakoutLevel);
    if (found >= 0) anchorIdx = found;
  }
  const anchored = bars.slice(anchorIdx);
  const anchoredDen = anchored.reduce((a, b) => a + b.volume, 0);
  const anchoredVwap = anchoredDen
    ? anchored.reduce((a, b) => a + ((b.high + b.low + b.close) / 3) * b.volume, 0) / anchoredDen
    : null;
  const aboveAnchoredVwap = anchoredVwap !== null && latest.close >= anchoredVwap;

  const last5 = mean(bars.slice(-5).map((b) => b.volume));
  const last20 = mean(bars.slice(-20).map((b) => b.volume));
  const dryUp = bars.length >= 11 && last5 < last20 * 0.85;

  const penaltyFlags: string[] = [];
  if (volumeRatio < 1.2) penaltyFlags.push("breakout_without_participation");
  const spread = latest.high - latest.low;
  const atrValue = last(atr(bars, 14), 0) || 0;
  if (atrValue && spread > atrValue * 1.5 && volumeRatio < 1.3) penaltyFlags.push("wide_spread_weak_volume");
  if (volumeRatio > 3 && latest.close < latest.high * 0.98) penaltyFlags.push("possible_exhaustion");

  return {
    volumeRatio: round(volumeRatio, 2),
    relativeVolume: round(volumeRatio, 2),
    obvConfirmation,
    vwapAlignment,
    anchoredVwap: anchoredVwap !== null ? round(anchoredVwap, 4) : null,
    aboveAnchoredVwap,
    dryUpBeforeExpansion: dryUp,
    accumulationTrend,
    penaltyFlags,
  };
}

function alignClose(asset: Bar[], benchmark: Bar[]): { a: number[]; b: number[] } {
  const benchMap = new Map<number, number>();
  for (const bar of benchmark) benchMap.set(bar.time, bar.close);
  const a: number[] = [];
  const b: number[] = [];
  for (const bar of asset) {
    const bc = benchMap.get(bar.time);
    if (bc !== undefined) {
      a.push(bar.close);
      b.push(bc);
    }
  }
  return { a, b };
}

export function analyzeRelativeStrength(
  bars: Bar[],
  benchmarkBars: Bar[],
  opts: { benchmarkSymbol: string; sectorBars?: Bar[] | null; sectorSymbol?: string | null },
): RelativeStrengthSignal {
  const { a, b } = alignClose(bars, benchmarkBars);
  if (!a.length) {
    return {
      benchmarkSymbol: opts.benchmarkSymbol,
      sectorBenchmarkSymbol: opts.sectorSymbol ?? null,
      score: 0,
      trendPersistence: 0,
      smoothness: 0,
      oneWeekAlpha: 0,
      oneMonthAlpha: 0,
      threeMonthAlpha: 0,
      sectorAlpha: null,
      explanation: "Benchmark overlap was unavailable.",
    };
  }

  const alpha = (window: number): number => {
    if (a.length < window + 1) return 0;
    const ar = a[a.length - 1] / a[a.length - 1 - window] - 1;
    const br = b[b.length - 1] / b[b.length - 1 - window] - 1;
    return ar - br;
  };

  const oneWeek = alpha(5);
  const oneMonth = alpha(21);
  const threeMonth = alpha(63);
  const rsLine = a.map((v, i) => v / b[i]);
  const smoothness = efficiencyRatio(rsLine, 20);
  let ups = 0;
  let total = 0;
  for (let i = Math.max(1, rsLine.length - 20); i < rsLine.length; i++) {
    if (rsLine[i] - rsLine[i - 1] > 0) ups++;
    total++;
  }
  const persistence = total ? ups / total : 0;

  let sectorAlpha: number | null = null;
  if (opts.sectorBars && opts.sectorBars.length) {
    const { a: sa, b: sb } = alignClose(bars, opts.sectorBars);
    if (sa.length > 22) {
      const ar = sa[sa.length - 1] / sa[sa.length - 22] - 1;
      const sr = sb[sb.length - 1] / sb[sb.length - 22] - 1;
      sectorAlpha = round((ar - sr) * 100, 2);
    }
  }

  const raw = (oneWeek * 20 + oneMonth * 35 + threeMonth * 45 + smoothness * 20 + persistence * 10) * 10;
  const score = Math.max(0, Math.min(100, raw));

  return {
    benchmarkSymbol: opts.benchmarkSymbol,
    sectorBenchmarkSymbol: opts.sectorSymbol ?? null,
    score: round(score, 2),
    trendPersistence: round(persistence * 100, 2),
    smoothness: round(smoothness * 100, 2),
    oneWeekAlpha: round(oneWeek * 100, 2),
    oneMonthAlpha: round(oneMonth * 100, 2),
    threeMonthAlpha: round(threeMonth * 100, 2),
    sectorAlpha,
    explanation:
      "Relative strength is measured against the selected benchmark with 1W/1M/3M alpha and path smoothness.",
  };
}

export function analyzeMarketRegime(benchmarkBars: Bar[], opts: { market: Market; benchmarkSymbol: string }): RegimeSignal {
  if (benchmarkBars.length < 220) {
    return {
      market: opts.market,
      benchmarkSymbol: opts.benchmarkSymbol,
      direction: "neutral",
      trendStrength: 0,
      volatilityState: "unknown",
      breadthLikeProxy: 0,
      explanation: "Insufficient benchmark history.",
    };
  }
  const c = closes(benchmarkBars);
  const ema20 = ema(c, 20);
  const ema50 = ema(c, 50);
  const ema200 = ema(c, 200);
  const atrSeries = atr(benchmarkBars, 14);
  const atrPctile = rollingPercentile(atrSeries, 252);
  const i = benchmarkBars.length - 1;
  const close = c[i];

  const bullish = close > ema20[i] && ema20[i] > ema50[i] && ema50[i] > ema200[i];
  const bearish = close < ema20[i] && ema20[i] < ema50[i] && ema50[i] < ema200[i];
  const direction: Direction = bullish ? "bullish" : bearish ? "bearish" : "neutral";

  const trendStrength =
    ([close > ema20[i], ema20[i] > ema50[i], ema50[i] > ema200[i]].map(Number).reduce((a, b) => a + b, 0) / 3) * 100;
  const atrP = Number.isFinite(atrPctile[i]) ? atrPctile[i] : 0.5;
  const volatilityState = atrP > 0.65 ? "expanding" : atrP < 0.35 ? "compressed" : "normal";

  let above = 0;
  const last50 = benchmarkBars.slice(-50);
  const startIdx = benchmarkBars.length - last50.length;
  for (let k = 0; k < last50.length; k++) above += last50[k].close > ema20[startIdx + k] ? 1 : 0;
  const breadthProxy = last50.length ? (above / last50.length) * 100 : 0;

  return {
    market: opts.market,
    benchmarkSymbol: opts.benchmarkSymbol,
    direction,
    trendStrength: round(trendStrength, 2),
    volatilityState,
    breadthLikeProxy: round(breadthProxy, 2),
    explanation: `${opts.market} regime is ${direction} with ${volatilityState} volatility relative to ${opts.benchmarkSymbol}.`,
  };
}
